src/ast/memory.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ast_memory_h
+#define wasm_ast_memory_h
+
+#include <algorithm>
+
+namespace wasm {
+
+namespace MemoryUtils {
+  // flattens memory into a single data segment. returns true if successful
+  inline bool flatten(Memory& memory) {
+    if (memory.segments.size() == 0) return true;
+    std::vector<char> data;
+    for (auto& segment : memory.segments) {
```

Comment:
Nit: at first glance I asked "why do we iterate over `memory.segments` twice?" An explanatory comment could be good, or even better extracted into another `MemoryUtil` function, `bool hasOnlyConstOffsets(Memory const&);`